File: 0166-fraction-to-recurring-decimal/0166-fraction-to-recurring-decimal.py
class Solution:
    def fractionToDecimal(self, numerator: int, denominator: int) -> str:
        sign = ""
        if numerator < 0 < denominator or denominator < 0 < numerator:
            sign += "-"
        
        numerator, denominator = abs(numerator), abs(denominator)
        whole = numerator // denominator
        rem = numerator % denominator
        if rem == 0:
            return sign + str(whole)
        
        dec = ""
        seen = {
            rem: 0
        }
        n = rem * 10
        while rem != 0:
            dec += str(n // denominator)
            rem = n % denominator
            if seen.get(rem) != None:
                break
            seen[rem] = len(dec)
            n = rem * 10
            
        if rem != 0:    
            start = seen.get(rem)
            dec = f"{dec[:start]}({dec[start:]})"
            
        return f"{sign}{whole}.{dec}"

File: 0166-fraction-to-recurring-decimal/0166-fraction-to-recurring-decimal.go
func fractionToDecimal(numerator int, denominator int) string {
    result := quotientSign(numerator, denominator)
    numerator, denominator = abs(numerator), abs(denominator)
    quotient, remainder := divmod(numerator, denominator)
    result += strconv.Itoa(quotient)
    
    if remainder == 0 {
        return result
    }
    
    decimal := ""
    seen := map[int]int{ remainder: 0 }
    for remainder != 0 {
        numerator = 10 * remainder
        quotient, remainder = divmod(numerator, denominator)
        decimal += strconv.Itoa(quotient)
        
        if _, ok := seen[remainder]; ok {
            break
        }
        seen[remainder] = len(decimal)
    }
    
    if remainder != 0 {
        repStart := seen[remainder]
        decimal = fmt.Sprintf("%s(%s)", decimal[:repStart], decimal[repStart:])
    }
    
    return fmt.Sprintf("%s.%s", result, decimal) 
}

func quotientSign(dividend, divisor int) string {
    if dividend < 0 && 0 < divisor ||
    divisor < 0 && 0 < dividend {
        return "-"
    } 
    return ""
}

func divmod(dividend, divisor int) (int, int) {
    quotient := dividend / divisor
    remainder := dividend % divisor
    return quotient, remainder
}

func abs(x int) int {
    if x < 0 {
        return -x
    }
    return x
}